hazelcast/src/main/java/com/hazelcast/cache/impl/CacheEventListenerAdaptor.java

```diff
@@ -187,12 +187,19 @@ private void handleEvent(int type, Collection<CacheEventData> keys) {
     private Iterable<CacheEntryEvent<? extends K, ? extends V>> createCacheEntryEvent(Collection<CacheEventData> keys) {
         HashSet<CacheEntryEvent<? extends K, ? extends V>> evt = new HashSet<CacheEntryEvent<? extends K, ? extends V>>();
         for (CacheEventData cacheEventData : keys) {
-            final EventType eventType = CacheEventType.convertToEventType(cacheEventData.getCacheEventType());
-            final K key = toObject(cacheEventData.getDataKey());
-            final V newValue = toObject(cacheEventData.getDataValue());
-            final V oldValue;
+            EventType eventType = CacheEventType.convertToEventType(cacheEventData.getCacheEventType());
+            boolean isRemovedOrExpired = eventType == EventType.REMOVED || eventType == EventType.EXPIRED;
+            K key = toObject(cacheEventData.getDataKey());
+            // when !isOldValueRequired and event type is removed or expired, both new & old value can be null
+            V newValue = (isRemovedOrExpired && !isOldValueRequired) ? null : (V) toObject(cacheEventData.getDataValue());
+            V oldValue;
             if (isOldValueRequired) {
-                oldValue = toObject(cacheEventData.getDataOldValue());
+                if (isRemovedOrExpired) {
+                    // according to contract of CacheEntryEvent#getValue
+                    oldValue = newValue;
+                } else {
+                    oldValue = toObject(cacheEventData.getDataOldValue());
+                }
```

Comment:
could the logic here be easier to follow when reading code if we re-write this part like this?
```
            boolean hasNewValue = !(eventType == EventType.REMOVED || eventType == EventType.EXPIRED);
            V newValue;
            V oldValue;
            if (isOldValueRequired) {
                if (hasNewValue) {
                    newValue = toObject(cacheEventData.getDataValue());
                    oldValue = toObject(cacheEventData.getDataOldValue());
                } else {
                    // according to contract of CacheEntryEvent#getValue
                    oldValue = toObject(cacheEventData.getDataValue());
                    newValue = oldValue;
                }
            } else {
                if (hasNewValue) {
                    newValue = toObject(cacheEventData.getDataValue());
                    oldValue = null;
                } else {
                    newValue = null;
                    oldValue = null;
                }
            }
```

Reply:
Done